DSS API specification cleanup & update

bskinner opened the pull request:
## Context

* Cleanup of discriminator use
* Cleanup of some date/time format descriptions
* Adds more `format` and `description` elements
* Removes `speccy` and switches to [redocly/openapi-cli](https://github.com/Redocly/openapi-cli)
* Updates summaries to fix linter errors, and prevent operation ids from appearing in the index column
* Removed unreferenced types & schemas

![Mailing Address Before & After](https://user-images.githubusercontent.com/274456/121087446-78386500-c7b2-11eb-8549-abcbee134237.png)


Cc. @yufengwng @ssettipalli 

## FUD Score

- [x] :relaxed: All good, business as usual

## How do we demo these changes?

in `docs/specification` run `npm run start`. The docs can then be accessed at `http://127.0.0.1:8080`

## Testing

- [x] I have written zero automated tests but have poked around locally to verify proper functionality

## Release

- [x] These changes require no special release procedures--just code!

MocanaAtBroad commented on pepper-apis/docs/specification/src/components/schemas.yml:
```diff
@@ -57,15 +59,17 @@ JoinMailingListPayload:
       description: The studyGuid of the mailing list the user is interested in (one of umbrellaId OR studyGuid but not both is required)
       example: ANGIO
       type: string
+      format: guid
```
Is this `guid` format something that is defined somewhere (in this doc or is a well-known thing)?

bskinner replied:
Nope, it's not- the list of supported formats are defined [here](https://swagger.io/specification/#data-type-format). If the value of the `format` property isn't one of the known values, tools are supposed to treat it as a open-string valued property that can have any value.

From my understanding of the platform, the definition of `GUID` is pulling double-duty- as a specific set of types (Study GUID, Answer GUID, etc), and also as a general way of identifying a unique identifier type. This is using the latter definition, and my intent here is to more visibly call out the identifier types in the API outside of just their name.

MocanaAtBroad reviewed (commented):
I see why you wanted to upgrade! A bunch of stuff got cleaned up, particularly those refs. discriminators are also a nice addition.
Looks good overall and redoc and related scripts seem to handle new spec just fine.
Have one question.

bskinner commented:
It feels quite nice 😂- those `$refs` and the `discriminator` weirdness have been bugging me from the beginning.

> I see why you wanted to upgrade! A bunch of stuff got cleaned up, particularly those refs. discriminators are also a nice addition.
> Looks good overall and redoc and related scripts seem to handle new spec just fine.
> Have one question.

bskinner committed:
Typo fixes

fixup: typos

fixup: typo

bskinner committed:
$refs ignore sibling elements

bskinner committed:
Adds format and descriptions to some schemas

bskinner committed:
Match the activity detail readonly usage

bskinner committed:
Add a couple safety-net defaults

bskinner committed:
Date/Time format definitions

* The OpenAPI spec defines the `date-time` by referencing RFC3339, section 5.6. The DSS API, however, includes milliseconds in it's date/times which is not part of the Internet Date/Time format.

* The DSS API does use an RFC3339 compatible Date format so the OpenAPI `date` format can be used in certain situations.

bskinner committed:
Explicitly describe the datetime format string

bskinner committed:
Remove messy $ref links

openapi-cli has a reasonable search path when resolving $ref paths, which means we can use local references here!

bskinner committed:
discriminator cleanup

bskinner committed:
Removes unused types

bskinner committed:
Linter fixes

bskinner committed:
Adds operation summaries

In some cases, the description is short enough that it would be better suited as a summary given the differences in presentation.

bskinner committed:
Remove speccy

bskinner commented:
This is ready for review!

bskinner committed:
Update join.mailing.list.yml

yufengwng commented on pepper-apis/docs/specification/src/endpoints/user.profile.address.temp.yml:
```diff
@@ -39,7 +39,7 @@ put:
 delete:
   operationId: deleteTemporaryMailingAddress
-  description: removed the temporary mailing address for a participant
+  summary: Delete a patient's temporary mailing address
```
```suggestion
  summary: Delete a participant's temporary mailing address
```

yufengwng commented on pepper-apis/docs/specification/src/components/schemas.yml:
```diff
@@ -2117,24 +2171,27 @@ Invitation:
       type: string
       nullable: true
       example: 2020-05-13T19:58:33.887Z
+      format: YYYY-MM-DDThhmmss.sssZ (ISO-8601)
```
Do we need colon separator in the time portion of the format?

bskinner replied:
Ooo, good catch. Yes, we do- we're using the extended format, not the basic. Thank you!

bskinner replied:
Fixed in 829b72f48950963b66cc03e82b6b03020ae5fe80

yufengwng reviewed (commented):
Looks great!

bskinner committed:
Wording correction

Co-authored-by: Yufeng Wang <yufeng@broadinstitute.org>

bskinner committed:
Adds colons to datetime format

yufengwng reviewed (dismissed):
👍

bskinner commented:
Brought the branch up to date with `develop`.
